### PR TITLE
[AAP-15916] Update error handling on PageForm

### PR DESCRIPTION
**Pull request by nixocio**
Create genericErrorAdapter as default error handling on PageForm.

Create awxErrorAdapater to deal with AWX error edge cases.

Create hubErrorAdapter to deal with Hub error edge cases.

Refactor RequestError and add unit tests.

Create a wrapper around PageForm default errorAdapter to be awxErrorAdapter.
This one should be used as part of the AWX project.

Create a wrapper around PageForm default errorAdapter to be hubErrorAdapter.
This one should be used as part of the Hub project.

Remove getAwxError in favor or error handling inside PageForm - or its generic version.

Modify CI to run requestError tests.

image

### CODE REVIEW

File: framework/PageForm/PageForm.tsx [hunk 1]
```diff
@@ -215,11 +217,17 @@ export function PageFormGrid(props: {
 
 export function PageFormSubmitButton(props: { children: ReactNode; style?: CSSProperties }) {
   const { isSubmitting, errors } = useFormState();
+  const { clearErrors } = useForm();
   const { t } = useTranslation();
   const hasErrors = errors && Object.keys(errors).length > 0;
   return (
     <Tooltip content={t('Please fix errors')} trigger={hasErrors ? undefined : 'manual'}>
       <Button
+        onClick={() => {
```
Comment (nixocio) [inline framework/PageForm/PageForm.tsx:226]: It was decided to clean Errors in case we have errors that not specific per field. Thus when the form is submitted the second time the error will be displayed again if remains the same.

**Comment by nixocio** [comment]
I found one more edge case on the AWX adapter, we will have to deal with objects like
 ```json
{
	"__all__": [
		"Inventory with this Name and Organization already exists."
	]
}
```

**Review by AlexSCorey — COMMENTED** [review]
I'm not quite done looking at this, but looks great so far.

File: cypress.common.config.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+import { defineConfig } from 'cypress';
+import { baseConfig } from './cypress.base.config';
+
+
+baseConfig.component!.specPattern = 'frontend/common/**/*.cy.{js,jsx,ts,tsx}';
```
[the remark this answers is not in the document]
Comment (nixocio) [reply cypress.common.config.ts:6]: Right now we do not have, I just followed the standard from the remaining files.

File: framework/PageForm/genericErrorAdapter.cy.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  it('should return generic error when error is an instance of Error', () => {
+    const error = new Error('Test Error');
+    const result = genericErrorAdapter(error);
+    expect(result).to.deep.equal({
+      genericErrors: [{ name: '__nonfield', message: 'Test Error' }],
+      fieldErrors: [],
+    });
+  });
+
+  it('should return empty errors when error is not a known error', () => {
```
[the remark this answers is not in the document]
Comment (nixocio) [reply framework/PageForm/genericErrorAdapter.cy.ts:14]: I will test that one a bit more, but an unknown error should be deal as a form level error.

File: framework/PageForm/PageForm.tsx [hunk 4]
```diff
@@ -133,12 +135,12 @@ export function PageForm<T extends object>(props: {
         {error && (
           <Alert
             variant="danger"
-            title={error.message ?? ''}
+            title={t('Error')}
```
Comment (himdel) [inline framework/PageForm/PageForm.tsx:138]: is there no support for message title and message description now?

Comment (nixocio) [reply framework/PageForm/PageForm.tsx:138]: @himdel I can add a conditional to add this one, but since error now is a string, or list of string messages, I refactored that one.

**Comment by himdel** [comment]
So.. the interface between the backend-specific parts and the shared parts is an object like...

```js
{
  genericErrors: [
    { name: '_nonFieldError', message: 'Test Error' },
  ],
  fieldErrors: [
    { name: 'name', message: 'Invalid name' },
  ],
}
```

The `fieldErrors` part looks good :+1:  (maybe `{ [fieldName]: [messages] }` would be easier to deal with than `[ {fieldName, message} ]` but sounds like a detail).

But .. the `genericErrors` part .. this bit should use the same kind of logic as the generic error handling elsewhere,
I know it's slightly out of scope for this PR, but:

* `name: '_nonFieldError',` - since this is not associated with any fields, it shouldn't pretend to have the field info .. IMO this just comes from the galaxy API hack where we needed to fit generic errors into an existing `{field:string}` format.
* `message` - most generic errors have either a single string, or a title/description pair .. we should support both?

**Comment by himdel** [comment]
Actually, one more bit .. there were some parts of the UI where we needed the error message to be jsx, not just a string.
(Think `Successfully saved <b>Whatever</b>`)

So .. is the errorAdapter the right place to produce that jsx (it's probably the last place that sees the raw data)?
(I'm not seeing the errorAdapter interface and types documented in the README, so maybe I'm missing something? .. but most of the code deals with type string, not ReactNode)

File: framework/PageForm/PageForm.tsx (hunk 1 above)
Comment (marshmalien) [reply framework/PageForm/PageForm.tsx:226]: This is a great solution to clearing form-level errors.

File: frontend/awx/adapters/awxErrorAdapter.tsx [hunk 5]
```diff
@@ -0,0 +1,40 @@
+import { ErrorOutput, GenericError } from '../../../framework/PageForm/typesErrorAdapter';
+import { isRequestError } from '../../common/crud/RequestError';
+
+export const awxErrorAdapter = (error: unknown): ErrorOutput => {
+  const genericErrors: GenericError[] = [];
+  const fieldErrors: GenericError[] = [];
+
+  if (
```
Comment (kialam) [inline frontend/awx/adapters/awxErrorAdapter.tsx:8]: This conditional feels overloaded. Is there a way to use a `Switch` statement that checks for `__all__ ` within `error.json` instead?

File: frontend/hub/adapters/hubErrorAdapter.tsx [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  title: string;
+}
+
+export const hubErrorAdapter: ErrorAdapter = (error): ErrorOutput => {
+  // errors can come in several flavors depending on if the API is from
+  // pulp or ansible.
+  const genericErrors: GenericError[] = [];
+  const fieldErrors: GenericError[] = [];
+
+  if (
```
Comment (kialam) [inline frontend/hub/adapters/hubErrorAdapter.tsx:22]: Can there be a `ResponseError` interface that checks for these attributes rather than having this evaluated with a set of conditionals?

File: framework/PageForm/PageForm.tsx (hunk 4 above)
Comment (nixocio) [reply framework/PageForm/PageForm.tsx]: I moved the Alert to its own component now, allowing different titles and components, based on the type of error.

File: frontend/hub/adapters/hubErrorAdapter.tsx (hunk 6 above)
Comment (nixocio) [reply frontend/hub/adapters/hubErrorAdapter.tsx]: Thanks for your review @kialam, I am refactoring those conditions.

**Review by himdel — APPROVED** [review]
Looks good :+1:, it *should* handle all the errors I know of :crossed_fingers: :)

(I haven't *thoroughly* checked the awx parts)

**Review by jamestalton — APPROVED** [review]
lgtm

**Comment by nixocio** [comment]
I had to rebase this one a couple of times to fix conflicts. Hopefully, last one.

File: framework/PageForm/genericErrorAdapter.cy.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (nixocio) [reply framework/PageForm/genericErrorAdapter.cy.ts:14]: If an instance of native Error will be treated as form level error.

**Review by marshmalien — APPROVED** [review]
Nice work! The changes look good to me!
